### PR TITLE
Avoid upcasting of subclasses of Symbol in IndexedBase.__new__

### LINKED ISSUE #17652: __new__ of IndexedBase upcasts subclass of Symbol to Symbol
First, thanks for creating SymPy. We're using it in a code generation framework: https://github.com/mabau/pystencils

We noticed that our tests are failling when using the current master of SymPy. To reproduce run:

```bash
git clone https://github.com/mabau/pystencils
cd pystencils
pip install -e . 
pip install --upgrade --ignore-installed git+https://github.com/sympy/sympy.git
python setup.py quicktest
``` 
While

```bash
git clone https://github.com/mabau/pystencils
cd pystencils
pip install -e . 
pip install --upgrade --ignore-installed git+https://github.com/sympy/sympy.git@sympy-1.4
python setup.py quicktest
```
should work fine.

The reason for this failure seems to be in the call of the constructor of  `sympy.IndexedBase`. We're a calling it with a subclass of  `sympy.Symbol` (`pystencils.data_types.TypedSymbol`) which is essentially a sympy Symbol with type information.

Is there a reason why the out-commented line is necessary? It causes our TypeSymbol to lose its type information by re-constructing a new Symbol. When I out-comment this line everything works fine again. 

In sympy/tensor/indexed.py
```python
    def __new__(cls, label, shape=None, **kw_args):
        assumptions, kw_args = _filter_assumptions(kw_args)
        if isinstance(label, string_types):
            label = Symbol(label)
        elif isinstance(label, Symbol):
            assumptions = label._merge(assumptions)
            # label = Symbol(label.name)
        elif isinstance(label, (MatrixBase, NDimArray)):
            return label
        elif isinstance(label, Iterable):
            return _sympify(label)
        else:
            label = _sympify(label)

        if is_sequence(shape):
            shape = Tuple(*shape)
        elif shape is not None:
            shape = Tuple(shape)

        offset = kw_args.pop('offset', S.Zero)
        strides = kw_args.pop('strides', None)

        if shape is not None:
            obj = Expr.__new__(cls, label, shape)
        else:
            obj = Expr.__new__(cls, label)
        obj._shape = shape
        obj._offset = offset
        obj._strides = strides
        obj._name = str(label)

        IndexedBase._set_assumptions(obj, assumptions)
        return obj
```

@mabau

### PR DESCRIPTION
**Pull request by theHamsta**
#### References to other Issues or PRs
Fixes #17652

#### Brief description of what is fixed or changed
Motivation is described in #17652.

If you call `a = IndexedBase(label)` where label is an instance of a subclass of  `sympy.Symbol` the type of `a.label` will be `sympy.Symbol` instead of the subclass.

Reason (this line in IndexedBase.__new__):
```python
label = Symbol(label.name)
```
that deletes all additional information that label might have as a subclass and also the assumptions that have been derived from it the line before. So the this test needs to should be adapted:

```
def test_IndexedBase_assumptions_inheritance():
    I = Symbol('I', integer=True)
    I_inherit = IndexedBase(I)
    I_explicit = IndexedBase('I', integer=True)

    assert I_inherit.is_integer
    assert I_explicit.is_integer
    assert I_inherit.label.is_integer # add this line ,previously None
    assert I_explicit.label.is_integer # add this line, previously None
    assert I_inherit == I_explicit
```

If you just delete the line `label = Symbol(label.name)`  #17652 would also be fixed. 
However,  `assert I_inherit == I_explicit` will no longer hold since `I_inherit.label.is_integer == True` while  `I_explicit.label.is_integer == None`.

#### Other comments

For me, it is not clear whether the constructor of `IndexedBase` should transfer the assumptions it generated back to `IndexedBase.label`. However, this seems more logical for me given the statement *the same assumptions for IndexedBase and IndexedBase.label (the symbol it was constructed from should hold*

#### Release Notes

* tensor
   * Allow construction of `IndexedBase` with subclasses of `Symbol` without upcasting to `Symbol`

### CODE REVIEW

**Comment by theHamsta** [comment]
Will try to fix the tests.

File: sympy/tensor/tests/test_indexed.py [hunk 1]
```diff
@@ -229,7 +229,21 @@ def test_IndexedBase_assumptions_inheritance():
 
     assert I_inherit.is_integer
     assert I_explicit.is_integer
-    assert I_inherit == I_explicit
+    assert I_inherit.label.is_integer
+    assert I_explicit.label.is_integer
+    assert I_explicit == I_inherit
+
+
+def test_IndexedBase_no_upcasting_of_label():
```
Comment (czgdp1807) [inline sympy/tensor/tests/test_indexed.py]: A better name can be `def test_issue_17652`.

**Comment by czgdp1807** [comment]
@theHamsta Please add a release notes entry. @Upabjojr Please review this PR.

**Comment by theHamsta** [comment]
I am still unsure what is the sympy way of transfering assumptions to the newly constructed `label` from string. `IndexedBase._set_assumptions(label, assumptions)` is kind of a hack.

Maybe you could help me with that.

I guess `label = Symbol(label.name)` was originally added to make the assertion assert `I_inherit == I_explicit` in  `test_IndexedBase_assumptions_inheritance`.

File: sympy/tensor/tests/test_indexed.py (hunk 1 above)
Comment (theHamsta) [reply sympy/tensor/tests/test_indexed.py]: :+1:

**Comment by Upabjojr** [comment]
Maybe `__new__` should support the `**assumptions` argument.

**Comment by Upabjojr** [comment]
Or maybe a dictionary merge?

**Comment by czgdp1807** [comment]
@Upabjojr Any comments on the latest changes? 
I will merge this if no objections are raised for 24 hours and the tests are passing.

File: sympy/tensor/indexed.py [hunk 2]
```diff
@@ -437,10 +437,9 @@ def __new__(cls, label, shape=None, **kw_args):
 
         assumptions, kw_args = _filter_assumptions(kw_args)
         if isinstance(label, string_types):
-            label = Symbol(label)
+            label = Symbol(label, **assumptions)
         elif isinstance(label, Symbol):
             assumptions = label._merge(assumptions)
-            label = Symbol(label.name)
```
Comment (Upabjojr) [inline sympy/tensor/indexed.py]: What about if you specify additional assumptions in `IndexedBase` that are not in the label? Is `_merge` also adding the additional assumptions to `label`?

**Comment by Upabjojr** [comment]
> 
> 
> @Upabjojr Any comments on the latest changes?
> I will merge this if no objections are raised for 24 hours and the tests are passing.

Just that small doubt, otherwise I'm OK for merging.

**Comment by theHamsta** [comment]
`_merge` does not change label. So additional assumptions from IndexedBased do not affect the label.

e.g. the last check of this test would fail

```python
def test_IndexedBase_assumptions_inheritance():
    I = Symbol('I', integer=True)
    I_inherit = IndexedBase(I)
    I_explicit = IndexedBase('I', integer=True)

    assert I_inherit.is_integer
    assert I_explicit.is_integer
    assert I_inherit.label.is_integer
    assert I_explicit.label.is_integer
    assert I_inherit == I_explicit

    I_inherit_and_explicit = IndexedBase(I, positive=True)

    assert I_inherit_and_explicit.is_integer
    assert I_inherit_and_explicit.is_positive
    assert I_inherit_and_explicit.label.is_integer
    assert I_inherit_and_explicit.label.is_positive
```

Because of that,  I had to set the assumption manually back to label in my first version.

I think this is a philosophical question. Should the constructor of class `IndexedBase` only import assumptions and leave its arguments unchanged (as you would expect from most constructors) or should it also change its arguments for consistency.
I don't know how this is handled in the rest of the project.

Should I add `self.label._assumptions = assumptions`?

**Comment by Upabjojr** [comment]
Maybe better to just create a new label?

**Comment by Upabjojr** [comment]
I'm not actually sure about this assumptions replication. Ideally the assumptions should just remain in the external body.

Have you thought about subclassing IndexedBase instead?

**Comment by theHamsta** [comment]
I only added the assumption replication to let `assert I_inherit == I_explicit` pass which would not be necesarry in my opinion. I think `label=Symbol(label.name)` was only added to fulfil this check.

I would just expect that `IndexedBase.label` would be exactly what you passed into `IndexedBase` without any modifications as it is in the current state of the PR. `IndexedBase.label` should just be reference to the original Symbol. I wouldn't care too much about what assumptions this label fulfils.

**Comment by Upabjojr** [comment]
I'm +1 for merging.

**Comment by bjodah** [comment]
Thank you for working on this!

**Pull request merged**